File: src/main/java/dev/cernavskis/moose/lexer/Lexer.java
package dev.cernavskis.moose.lexer;

import java.util.*;

// Implements iterator so that we can implement this as a stream.
// This is done to save memory, as well as it allows us to run this on a seperate thread.
public class Lexer {
  private final String input;
  private int position = 0;
  // These next three are used for debugging and error messages.
  private int line = 1;
  private int column = 1;
  private final String file;

  public Lexer(String input) {
    this(input, "<unknown>");
  }

  public Lexer(String input, String file) {
    this.input = input;
    this.file = file;
  }

  private boolean hadEOF = false;
  public List<Token> getAllTokens() {
    List<Token> tokens = new LinkedList<>();
    while (!hadEOF) {
      Token token = nextToken();
      tokens.add(token);
      hadEOF = token.type() == TokenType.EOF;
    }
    return new ArrayList<>(tokens);
  }

  // Okay, this is the real deal.
  private Token nextToken() {
    // Loop over everything while we still have input.
    while (position < input.length()) {
      char currentChar = input.charAt(position);

      // Skip whitespace.
      if (currentChar == ' ' || currentChar == '\n' || currentChar == '\r' || currentChar == '\t') {
        incrPos();
        continue;
      }

      // Skip comments.
      if (currentChar == '/' && input.charAt(position + 1) == '/') {
        incrPos();
        incrPos();
        while (position < input.length() && input.charAt(position) != '\n') {
          incrPos();
        }
        continue;
      }

      // Tokenize everything that is a single character
      if (currentChar == '[') {
        incrPos();
        return newToken(TokenType.ARRAY_LEFT, "[");
      }

      if (currentChar == ']') {
        incrPos();
        return newToken(TokenType.ARRAY_RIGHT, "]");
      }

      if (currentChar == '(') {
        incrPos();
        return newToken(TokenType.PAREN_LEFT, "(");
      }
      if (currentChar == ')') {
        incrPos();
        return newToken(TokenType.PAREN_RIGHT, ")");
      }
      if (currentChar == '{') {
        incrPos();
        return newToken(TokenType.BLOCK_LEFT, "{");
      }
      if (currentChar == '}') {
        incrPos();
        return newToken(TokenType.BLOCK_RIGHT, "}");
      }
      if (currentChar == ',') {
        incrPos();
        return newToken(TokenType.COMMA, ",");
      }

      if (currentChar == ';') {
        incrPos();
        return newToken(TokenType.SEMICOLON, ";");
      }

      if (currentChar == '$') {
        incrPos();
        StringBuilder asm = new StringBuilder();
        while (position < input.length() && input.charAt(position) != '$') {
          asm.append(input.charAt(position));
          incrPos();
        }
        incrPos();
        return newToken(TokenType.ASM, asm.toString());
      }

      // More complex tokens down below, see the functions for more info.
      if (isOperatorChar(currentChar)) {
        String operator = getOperator();
        if (isValidOperator(operator)) {
          TokenType type;
          switch (operator) {
            case "=" -> type = TokenType.ASSIGNMENT;
            case "+" -> type = TokenType.ADDITION;
            case "-" -> type = TokenType.SUBTRACTION;
            case "*" -> type = TokenType.MULTIPLICATION;
            case "/" -> type = TokenType.DIVISION;
            case "+=" -> type = TokenType.ADDITION_ASSIGNMENT;
            case "-=" -> type = TokenType.SUBTRACTION_ASSIGNMENT;
            case "*=" -> type = TokenType.MULTIPLICATION_ASSIGNMENT;
            case "/=" -> type = TokenType.DIVISION_ASSIGNMENT;
            case "%" -> type = TokenType.MODULO;
            case "%=" -> type = TokenType.MODULO_ASSIGNMENT;
            case "++" -> type = TokenType.PREINCREMENT;
            case "--" -> type = TokenType.PREDECREMENT;
            case "**" -> type = TokenType.EXPONENTIATION;
            case "**=" -> type = TokenType.EXPONENTIATION_ASSIGNMENT;
            case "==" -> type = TokenType.EQ;
            case "!=" -> type = TokenType.NEQ;
            case ">=" -> type = TokenType.GTE;
            case "<=" -> type = TokenType.LTE;
            case ">" -> type = TokenType.GT;
            case "<" -> type = TokenType.LT;
            case "||" -> type = TokenType.LOGICAL_OR;
            case "&&" -> type = TokenType.LOGICAL_AND;
            case "!" -> type = TokenType.LOGICAL_NOT;
            case "." -> type = TokenType.DOT;
            case "?" -> type = TokenType.TERNARY;
            case ":" -> type = TokenType.COLON;
            case "~" -> type = TokenType.BIT_NOT;
            case "|" -> type = TokenType.BIT_OR;
            case "&" -> type = TokenType.BIT_AND;
            case "^" -> type = TokenType.BIT_XOR;
            case "~=" -> type = TokenType.BIT_NOT_ASSIGNMENT;
            case "|=" -> type = TokenType.BIT_OR_ASSIGNMENT;
            case "&=" -> type = TokenType.BIT_AND_ASSIGNMENT;
            case "^=" -> type = TokenType.BIT_XOR_ASSIGNMENT;
            case ">>" -> type = TokenType.BIT_RSHIFT;
            case "<<" -> type = TokenType.BIT_LSHIFT;
            case ">>=" -> type = TokenType.BIT_RSHIFT_ASSIGNMENT;
            case "<<=" -> type = TokenType.BIT_LSHIFT_ASSIGNMENT;
            default -> throw new RuntimeException("Invalid operator: " + operator); // Unreachable
          }

          return newToken(type, operator);
        } else {
          throw new IllegalArgumentException("Invalid operator: " + operator);
        }
      }

      if (isString(currentChar)) {
        String value = getString();
        return newToken(TokenType.STRING, value);
      }


      if (isIdentifier(currentChar)) {
        String value = getIdentifier();
        return switch (value) {
          case "func" -> newToken(TokenType.FUNC, value);
          case "let" -> newToken(TokenType.LET, value);
          case "const" -> newToken(TokenType.CONST, value);
          case "if" -> newToken(TokenType.IF, value);
          case "else" -> newToken(TokenType.ELSE, value);
          case "for" -> newToken(TokenType.FOR, value);
          case "foreach" -> newToken(TokenType.FOREACH, value);
          case "do" -> newToken(TokenType.DO, value);
          case "while" -> newToken(TokenType.WHILE, value);
          case "loop" -> newToken(TokenType.LOOP, value);
          case "return" -> newToken(TokenType.RETURN, value);
          case "break" -> newToken(TokenType.BREAK, value);
          case "continue" -> newToken(TokenType.CONTINUE, value);
          case "import" -> newToken(TokenType.IMPORT, value);
          case "export" -> newToken(TokenType.EXPORT, value);
          default -> newToken(TokenType.IDENTIFIER, value);
        };
      }

      if (isConstant(currentChar)) {
        String value = getConstant();
        return newToken(TokenType.CONSTANT, value);
      }

      incrPos();
    }
    return newToken(TokenType.EOF, "");
  }

  // Is the current character a character used in operators?
  private boolean isOperatorChar(char currentChar) {
    return currentChar == '~' || currentChar == '|' || currentChar == '&' || currentChar == '^' ||
      currentChar == '+' || currentChar == '-' || currentChar == '*' || currentChar == '/' ||
      currentChar == '=' || currentChar == '!' || currentChar == '>' || currentChar == '<' ||
      currentChar == '%' || currentChar == '?' || currentChar == ':' || currentChar == '.';
  }

  // Validate that the operator actually exists.
  private boolean isValidOperator(String operator) {
    return Arrays.asList(new String[]{
      // Bitwise
      "~", "|", "&", "^", "~=", "|=", "&=", "^=",
      ">>", "<<", ">>=", "<<=",
      // Numerical
      "+", "-", "*", "/", "+=", "-=", "*=", "/=",
      "%", "%=", "++", "--", "**", "**=",
      // Comparison
      ">", "<", "==", "!=", ">=", "<=",
      // Boolean
      "||", "&&", "!",
      // Miscellaneous
      "?", ":", ".", "=",
    }).contains(operator);
  }

  // Loop over everything until we no longer have an operator character
  private String getOperator() {
    StringBuilder builder = new StringBuilder();
    while (position < input.length() && isOperatorChar(input.charAt(position))) {
      builder.append(input.charAt(position));
      incrPos();
    }
    return builder.toString();
  }

  // Ooooh do we have the start of a string?
  private boolean isString(char currentChar) {
    return currentChar == '"';
  }

  // Basic JSON-type string parsing.
  // TODO: uXXXX and xXX escaping support.
  private String getString() {
    StringBuilder sb = new StringBuilder();
    incrPos();
    while (position < input.length() && input.charAt(position) != '"') {
      if (input.charAt(position) == '\\') {
        incrPos();
        if (position < input.length()) {
          if (input.charAt(position) == 'n') {
            sb.append('\n');
          } else if (input.charAt(position) == 't') {
            sb.append('\t');
          } else if (input.charAt(position) == '"') {
            sb.append('"');
          } else if (input.charAt(position) == '\\') {
            sb.append('\\');
          } else {
            sb.append(input.charAt(position));
          }
        }
      } else {
        sb.append(input.charAt(position));
      }
      incrPos();
    }
    incrPos();
    return sb.toString();
  }

  // Is this the start of an identifier? Identifiers are alphanumeric, and can contain underscores. They must not start with a number, otherwise they are parsed as a constant.
  private boolean isIdentifier(char c) {
    return Character.isLetter(c) || c == '_';
  }

  // Here we check for a number in the input.
  private String getIdentifier() {
    StringBuilder sb = new StringBuilder();
    while (position < input.length() && (Character.isLetterOrDigit(input.charAt(position)) || input.charAt(position) == '_')) {
      sb.append(input.charAt(position));
      incrPos();
    }
    return sb.toString();
  }

  // Is this the start of a constant? Constants are just numbers. Constants can start with a number or a minus sign and a number after it.
  private boolean isConstant(char c) {
    return (Character.isDigit(c) || (c == '-' && Character.isDigit(input.charAt(position + 1))));
  }

  // Here we also check for a decimal point in the input.
  private String getConstant() {
    StringBuilder sb = new StringBuilder();
    while (position < input.length() && (Character.isDigit(input.charAt(position)) || input.charAt(position) == '.')) {
      sb.append(input.charAt(position));
      incrPos();
    }
    return sb.toString();
  }

  // Utility function to not need to write getDebugInfo every time.
  private Token newToken(TokenType type, String value) {
    return new Token(type, value, line, column, file);
  }

  // Analagous to position++, but also increments line and column correctly.
  private void incrPos() {
    if (input.charAt(position) == '\n') {
      line++;
      column = 0;
    } else {
      column++;
    }
    position++;
  }
}

File: src/lexer/lexer.rs
use super::{Token, TokenType};

pub struct Lexer {
    input: String,
    position: usize,
    line: usize,
    column: usize,
    file: String,
}

impl Lexer {
    pub fn new(input: &str, file: &str) -> Self {
        Lexer {
            input: input.to_string(),
            position: 0,
            line: 1,
            column: 1,
            file: file.to_string(),
        }
    }

    pub fn get_all_tokens(&mut self) -> Vec<Token> {
        let mut tokens = Vec::new();
        loop {
            let token = self.next_token();
            let is_eof = token.token_type == TokenType::Eof;
            tokens.push(token);
            if is_eof {
                break;
            }
        }
        tokens
    }

    fn next_token(&mut self) -> Token {
        while self.position < self.input.len() {
            let current_char = self.current_char();

            // Skip whitespace
            if current_char == ' ' || current_char == '\n' || current_char == '\r' || current_char == '\t' {
                self.incr_pos();
                continue;
            }

            // Skip comments
            if current_char == '/' && self.peek_char() == Some('/') {
                self.incr_pos();
                self.incr_pos();
                while self.position < self.input.len() && self.current_char() != '\n' {
                    self.incr_pos();
                }
                continue;
            }

            // Single character tokens
            if current_char == '[' {
                self.incr_pos();
                return self.new_token(TokenType::ArrayLeft, "[");
            }
            if current_char == ']' {
                self.incr_pos();
                return self.new_token(TokenType::ArrayRight, "]");
            }
            if current_char == '(' {
                self.incr_pos();
                return self.new_token(TokenType::ParenLeft, "(");
            }
            if current_char == ')' {
                self.incr_pos();
                return self.new_token(TokenType::ParenRight, ")");
            }
            if current_char == '{' {
                self.incr_pos();
                return self.new_token(TokenType::BlockLeft, "{");
            }
            if current_char == '}' {
                self.incr_pos();
                return self.new_token(TokenType::BlockRight, "}");
            }
            if current_char == ',' {
                self.incr_pos();
                return self.new_token(TokenType::Comma, ",");
            }
            if current_char == ';' {
                self.incr_pos();
                return self.new_token(TokenType::Semicolon, ";");
            }

            // ASM token
            if current_char == '$' {
                self.incr_pos();
                let mut asm = String::new();
                while self.position < self.input.len() && self.current_char() != '$' {
                    asm.push(self.current_char());
                    self.incr_pos();
                }
                self.incr_pos();
                return self.new_token(TokenType::Asm, &asm);
            }

            // Operators
            if self.is_operator_char(current_char) {
                let operator = self.get_operator();
                if self.is_valid_operator(&operator) {
                    let token_type = match operator.as_str() {
                        "=" => TokenType::Assignment,
                        "+" => TokenType::Addition,
                        "-" => TokenType::Subtraction,
                        "*" => TokenType::Multiplication,
                        "/" => TokenType::Division,
                        "+=" => TokenType::AdditionAssignment,
                        "-=" => TokenType::SubtractionAssignment,
                        "*=" => TokenType::MultiplicationAssignment,
                        "/=" => TokenType::DivisionAssignment,
                        "%" => TokenType::Modulo,
                        "%=" => TokenType::ModuloAssignment,
                        "++" => TokenType::PreIncrement,
                        "--" => TokenType::PreDecrement,
                        "**" => TokenType::Exponentiation,
                        "**=" => TokenType::ExponentiationAssignment,
                        "==" => TokenType::Eq,
                        "!=" => TokenType::Neq,
                        ">=" => TokenType::Gte,
                        "<=" => TokenType::Lte,
                        ">" => TokenType::Gt,
                        "<" => TokenType::Lt,
                        "||" => TokenType::LogicalOr,
                        "&&" => TokenType::LogicalAnd,
                        "!" => TokenType::LogicalNot,
                        "." => TokenType::Dot,
                        "?" => TokenType::Ternary,
                        ":" => TokenType::Colon,
                        "~" => TokenType::BitNot,
                        "|" => TokenType::BitOr,
                        "&" => TokenType::BitAnd,
                        "^" => TokenType::BitXor,
                        "~=" => TokenType::BitNotAssignment,
                        "|=" => TokenType::BitOrAssignment,
                        "&=" => TokenType::BitAndAssignment,
                        "^=" => TokenType::BitXorAssignment,
                        ">>" => TokenType::BitRshift,
                        "<<" => TokenType::BitLshift,
                        ">>=" => TokenType::BitRshiftAssignment,
                        "<<=" => TokenType::BitLshiftAssignment,
                        _ => panic!("Invalid operator: {}", operator),
                    };
                    return self.new_token(token_type, &operator);
                } else {
                    panic!("Invalid operator: {}", operator);
                }
            }

            // Strings
            if current_char == '"' {
                let value = self.get_string();
                return self.new_token(TokenType::String, &value);
            }

            // Identifiers and keywords
            if self.is_identifier(current_char) {
                let value = self.get_identifier();
                let token_type = match value.as_str() {
                    "func" => TokenType::Func,
                    "let" => TokenType::Let,
                    "const" => TokenType::Const,
                    "if" => TokenType::If,
                    "else" => TokenType::Else,
                    "for" => TokenType::For,
                    "foreach" => TokenType::Foreach,
                    "do" => TokenType::Do,
                    "while" => TokenType::While,
                    "loop" => TokenType::Loop,
                    "return" => TokenType::Return,
                    "break" => TokenType::Break,
                    "continue" => TokenType::Continue,
                    "import" => TokenType::Import,
                    "export" => TokenType::Export,
                    _ => TokenType::Identifier,
                };
                return self.new_token(token_type, &value);
            }

            // Constants (numbers)
            if self.is_constant(current_char) {
                let value = self.get_constant();
                return self.new_token(TokenType::Constant, &value);
            }

            self.incr_pos();
        }
        
        self.new_token(TokenType::Eof, "")
    }

    fn current_char(&self) -> char {
        self.input.chars().nth(self.position).unwrap()
    }

    fn peek_char(&self) -> Option<char> {
        self.input.chars().nth(self.position + 1)
    }

    fn is_operator_char(&self, c: char) -> bool {
        matches!(c, '~' | '|' | '&' | '^' | '+' | '-' | '*' | '/' | '=' | '!' | '>' | '<' | '%' | '?' | ':' | '.')
    }

    fn is_valid_operator(&self, operator: &str) -> bool {
        matches!(operator,
            "~" | "|" | "&" | "^" | "~=" | "|=" | "&=" | "^=" |
            ">>" | "<<" | ">>=" | "<<=" |
            "+" | "-" | "*" | "/" | "+=" | "-=" | "*=" | "/=" |
            "%" | "%=" | "++" | "--" | "**" | "**=" |
            ">" | "<" | "==" | "!=" | ">=" | "<=" |
            "||" | "&&" | "!" |
            "?" | ":" | "." | "="
        )
    }

    fn get_operator(&mut self) -> String {
        let mut builder = String::new();
        while self.position < self.input.len() && self.is_operator_char(self.current_char()) {
            builder.push(self.current_char());
            self.incr_pos();
        }
        builder
    }

    fn get_string(&mut self) -> String {
        let mut result = String::new();
        self.incr_pos(); // Skip opening quote
        
        while self.position < self.input.len() && self.current_char() != '"' {
            if self.current_char() == '\\' {
                self.incr_pos();
                if self.position < self.input.len() {
                    match self.current_char() {
                        'n' => result.push('\n'),
                        't' => result.push('\t'),
                        '"' => result.push('"'),
                        '\\' => result.push('\\'),
                        c => result.push(c),
                    }
                }
            } else {
                result.push(self.current_char());
            }
            self.incr_pos();
        }
        
        self.incr_pos(); // Skip closing quote
        result
    }

    fn is_identifier(&self, c: char) -> bool {
        c.is_alphabetic() || c == '_'
    }

    fn get_identifier(&mut self) -> String {
        let mut result = String::new();
        while self.position < self.input.len() {
            let c = self.current_char();
            if c.is_alphanumeric() || c == '_' {
                result.push(c);
                self.incr_pos();
            } else {
                break;
            }
        }
        result
    }

    fn is_constant(&self, c: char) -> bool {
        c.is_numeric() || (c == '-' && self.peek_char().map_or(false, |ch| ch.is_numeric()))
    }

    fn get_constant(&mut self) -> String {
        let mut result = String::new();
        while self.position < self.input.len() {
            let c = self.current_char();
            if c.is_numeric() || c == '.' {
                result.push(c);
                self.incr_pos();
            } else {
                break;
            }
        }
        result
    }

    fn new_token(&self, token_type: TokenType, value: &str) -> Token {
        Token::new(token_type, value.to_string(), self.line, self.column, self.file.clone())
    }

    fn incr_pos(&mut self) {
        if self.position < self.input.len() && self.current_char() == '\n' {
            self.line += 1;
            self.column = 0;
        } else {
            self.column += 1;
        }
        self.position += 1;
    }
}
